scripts/speech_recognition/estimate_duration_bins_2d.py

```diff
@@ -0,0 +1,319 @@
+import math
+from functools import partial
+from itertools import islice
+from pathlib import Path
+from typing import Callable, Iterable
+
+import numpy as np
+import pandas as pd
+from lhotse.cut import Cut
+from omegaconf import OmegaConf
+
+from nemo.collections.asr.data.audio_to_text_lhotse import TokenizerWrapper
+from nemo.collections.common.data.lhotse.cutset import read_cutset_from_config
+from nemo.collections.common.data.lhotse.dataloader import (
+    DurationFilter,
+    FixedBucketBatchSizeConstraint2D,
+    LhotseDataLoadingConfig,
+    TokenPerSecondFilter,
+    tokenize,
+)
+from nemo.collections.common.prompts.formatter import PromptFormatter
+from nemo.collections.common.tokenizers import AggregateTokenizer, SentencePieceTokenizer
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="Estimate duration bins for Lhotse dynamic bucketing using a sample of the input dataset. "
+        "The dataset is read either from one or more manifest files and supports data weighting. "
+        "Unlike estimate_duration_bins.py, this script prepares the setup for 2D bucketing. "
+        "This means that each main bucket for audio duration is sub-divided into sub-buckets "
+        "for the number of output tokens (supporting BPE and Aggregated tokenizers). "
+        "2D bucketing is especially useful for encoder-decoder models where input audio duration is often "
+        "not sufficient to stratify the sampling with an optimal GPU utilization.",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+    )
+    parser.add_argument(
+        "input",
+        help='Data input. Options: '
+        '1) "path.json" - any single NeMo manifest; '
+        '2) "[[path1.json],[path2.json],...]" - any collection of NeMo manifests; '
+        '3) "[[path1.json,weight1],[path2.json,weight2],...]" - any collection of weighted NeMo manifests; '
+        '4) "input_cfg.yaml" - a new option supporting input configs, same as in model training \'input_cfg\' arg; '
+        '5) "path/to/shar_data" - a path to Lhotse Shar data directory; '
+        '6) "key=val" - in case none of the previous variants cover your case: "key" is the key you\'d use in NeMo training config with its corresponding value ',
+    )
+    parser.add_argument(
+        "-t",
+        "--tokenizer",
+        nargs="+",
+        required=True,
+        help="Path to one or more SPE tokenizers. More than one means we'll use AggregateTokenizer and --langs argument must also be used. When provided, we'll estimate a 2D distribution for input and output sequence lengths.",
+    )
+    parser.add_argument(
+        "-a", "--langs", nargs="+", help="Language names for each of AggregateTokenizer sub-tokenizers."
+    )
+    parser.add_argument(
+        "-b",
+        "--buckets",
+        type=int,
+        default=30,
+        help="The desired number of buckets (dim0 => covers input sequence length / audio duration).",
+    )
+    parser.add_argument(
```

Comment:
curious, is there a lot of difference between 150 1-d buckets vs 30x5 2-d buckets (after filtering out outliers)

Reply:
I don't think 150 1D buckets would help you a lot vs 30. With 30, I'm already seeing only 5% of input padding, but there is typically ~40% of output padding. With 2D buckets the exact improvement in padding ratios is not conclusive yet; for 30x2 I've seen it drop to ~23.5%, but with 30x5 I'm seeing ~31% (although there are visually a lot more batches that are close to ~5%, so the distribution seems more bi-modal). I don't really want to try 30x30 because that might take a lot of time to finish, partiallu defeating the purpose.

Anyway, I am positive that 30x2 is a better setup than 30x1 both in terms of throughput (60->80%) / batch-size (4x -> 5x) and a speed-up in convergence.